doc/sphinx/src/outputs.rst

```diff
@@ -158,6 +161,134 @@ This will produce a text file (``.hst``) output file every 1 units of
 simulation time. The content of the file is determined by the functions
 enrolled by a specific package, see :ref:`state history output`.
 
+Histograms
+----------
+
+Parthenon supports calculating flexible 1D and 2D histograms in-situ that
+are written to disk in HDF5 format.
+Currently supported are
+
+- 1D and 2D histograms
+- binning by variable or coordinate (x1, x2, x3 and radial distance)
+- counting samples and or summing a variable  
+- weighting by volume and/or variable
+
+The output format follows ``numpy`` convention, so that plotting data
+with Python based machinery should be straightfoward (see example below).
+In general, histograms are calculated using inclusive left bin edges and
+data equal to the rightmost edge is also included in the last bin.
+
+A ``<parthenon/output*>`` block containing one simple and one complex
+example might look like::
+
+   <parthenon/output8>
+   file_type = histogram # required, sets the output type
+   dt = 1.0              # required, sets the output interval
+   num_histograms = 2    # required, specifies how many histograms are defined in this block
+
+   # 1D histogram ("standard", i.e., counting occurance in bin)
+   hist0_ndim = 1
```

Comment:
The `hist0_`, `hist1_`, etc. prefixes are very ugly. Is it necessary to support more than 1 histogram per output file?

Reply:
I'm also not a big fan of this style, but it's the best I came up with so I'm open for suggestions (that do not involve moving to yaml/json/... for parsing inputs, which now seems to come up more often *wink* @bprather 

The main motivation for this approach is to not clobber the (shared) filesystem with many new files.
I can see that one easily calculate 10+ histograms in a simulation at high cadence, which would result in many tiny files (that'd need to be managed again).
To mitigate this, I initially even tried to put all histograms of all times in a single file, but gave up because of HDF5 (deleting/updating data cannot be easily done).

Reply:
Could always use yaml/json output to do that... is avoiding an extra library dependency the issue?

Reply:
This might be what you tried, but it is *supposed* to be possible to do streaming writes using HDF5 with the SWMR interface: https://docs.h5py.org/en/stable/swmr.html

Reply:
> To mitigate this, I initially even tried to put all histograms of all times in a single file, but gave up because of HDF5 (deleting/updating data cannot be easily done).

> This might be what you tried, but it is supposed to be possible to do streaming writes using HDF5 with the SWMR interface: https://docs.h5py.org/en/stable/swmr.html

I have in the past used extendable datasets to output all data at all times in a single file in HDF5. I do not recommend it. The main issue is that this strategy is not robust against failure. If the code crashes during an output, the entire file is corrupted, not just the time you're trying to output. It's also less robust against the kind of thing we often do, where we restart a simulation and steer it differently upon restart, for example. For these reasons I am a strong believer in 1 file per output time.

> I'm also not a big fan of this style, but it's the best I came up with so I'm open for suggestions

What about simply assuming one `output` block per histogram, but combining them into a single file if the `dt` is the same?

Reply:
Or alternatively have a `file_index` be a parameter per block, and all blocks with the same `file_index` are in the same file and assumed to have the same `dt`. If multiple blocks set `dt` this is either undefined behavior, or a runtime error.

Reply:
I gave this some more thought and am now sure how we should go ahead.
Originally, I liked @Yurlungur idea to have a default file_index and multiple output blocks as it looked clean.
Though when thinking about the implementation, I realized that this might introduce some arbitrary ordering in the resulting output file depending on the order of the `output#` blocks and number of different `file_index` given, which may result in additional fine tuning/logic when going from input file to postprocessing.

As ugly as the currently solution is, it's from that point of view unambiguous. All histograms in one output block are in one file and in the same order they are defined.
So I'm currently leaning towards keeping this verbose option but am also open to further suggestions (or going with the solution that @Yurlungur suggested if people prefer that).

Reply:
I'm obviously biased towards my own solution :P. I also think if one relies on ordering in an hdf5 file, one is asking for trouble.

That said, I don't feel that strongly and am fine to defer to implementers choice :)

Reply:
> I'm obviously biased towards my own solution :P. I also think if one relies on ordering in an hdf5 file, one is asking for trouble.

"Order" here referred to the explicit number used in the input file that is also being used as the hdf5 group name in the output file, so it's unique and does not depend on the literal order written down in the input file.

Having said that, it just occurred to me that we could enforce that each histogram has a unique string name (rather than a number) which is then used as prefix (both in the input file as well as the output file).

How about going from
```
num_histograms = 3

# 1D histogram of a variable, binned by a variable
hist0_ndim = 1
hist0_x_variable = advected
hist0_x_variable_component = 0
hist0_x_edges_type = list
hist0_x_edges_list = 1e-9, 1e-4,1e-1, 2e-1, 5e-1 ,1.0
hist0_binned_variable = advected
hist0_binned_variable_component = 0

# 2D histogram of a variable, binned by a coordinate and a different variable
hist1_ndim = 2
hist1_x_variable = HIST_COORD_X1
hist1_x_edges_type = lin
hist1_x_edges_num_bins = 4
...
```
to
```
hist_names = myname, rhoT

# 1D histogram of a variable, binned by a variable
myname_ndim = 1
myname_x_variable = advected
myname_x_variable_component = 0
myname_x_edges_type = list
myname_x_edges_list = 1e-9, 1e-4,1e-1, 2e-1, 5e-1 ,1.0
myname_binned_variable = advected
myname_binned_variable_component = 0

# 2D histogram of a variable, binned by a coordinate and a different variable
rhoT_ndim = 2
rhoT_x_variable = HIST_COORD_X1
rhoT_x_edges_type = lin
rhoT_x_edges_num_bins = 4
...
```

This (using names) would, of course, also work in the context of spreading the info over multiple output blocks (each with a unique name for a histogram), but I'm leaning towards keeping the current overall logic (one output block for one output file for a given dt) over adding additional content from different blocks.

What do people think?

Reply:
I like the new convention with names instead of indices. :+1: